rclrs/src/wait/guard_condition.rs

```diff
@@ -0,0 +1,176 @@
+/// by adding an `Arc<GuardCondition>` to the wait set, and calling `trigger()` on the same `GuardCondition` while
+/// the wait set is waiting.
+///
+/// The guard condition may be reused multiple times, but like other waitable entities, can not be used in
+/// multiple wait sets concurrently.
+///
+/// # Example
+/// ```
+/// # use rclrs::{Context, GuardCondition, WaitSet, RclrsError};
+/// # use std::sync::{Arc, atomic::Ordering};
+///
+/// let context = Context::new([])?;
+///
+/// let atomic_bool = Arc::new(std::sync::atomic::AtomicBool::new(false));
+/// let atomic_bool_for_closure = Arc::clone(&atomic_bool);
+///
+/// let gc = Arc::new(GuardCondition::new(
+///     &context,
+///     Some(Box::new(move || {
+///         atomic_bool_for_closure.store(true, Ordering::Relaxed);
+///     })),
+/// ));
+///
+/// let mut ws = WaitSet::new(0, 1, 0, 0, 0, 0, &context)?;
+/// ws.add_guard_condition(Arc::clone(&gc))?;
+///
+/// // Trigger the guard condition, firing the callback and waking the wait set being waited on, if any.
+/// gc.trigger()?;
+///
+/// // The provided callback has now been called.
+/// assert_eq!(atomic_bool.load(Ordering::Relaxed), true);
+///
+/// // The wait call will now immediately return.
+/// ws.wait(Some(std::time::Duration::from_millis(10)))?;
+///
+/// # Ok::<(), RclrsError>(())
+/// ```
+pub struct GuardCondition {
+    /// The rcl_guard_condition_t that this struct encapsulates.
+    pub(crate) rcl_guard_condition: Mutex<rcl_guard_condition_t>,
+    /// An optional callback to call when this guard condition is triggered.
+    callback: Option<Box<dyn Fn() + Send + Sync>>,
+    /// A flag to indicate if this guard condition has already been assigned to a wait set.
+    pub(crate) in_use_by_wait_set: Arc<AtomicBool>,
+}
+
+impl Drop for GuardCondition {
+    fn drop(&mut self) {
+        unsafe {
+            // SAFETY: No precondition for this function (besides passing in a valid guard condition)
+            rcl_guard_condition_fini(&mut *self.rcl_guard_condition.lock().unwrap());
+        }
+    }
+}
+
+impl PartialEq for GuardCondition {
+    fn eq(&self, other: &Self) -> bool {
+        // Because GuardCondition controls the creation of the rcl_guard_condition, each unique GuardCondition should have a unique
+        // rcl_guard_condition. Thus comparing equality of this member should be enough.
+        std::ptr::eq(
+            &self.rcl_guard_condition.lock().unwrap().impl_,
+            &other.rcl_guard_condition.lock().unwrap().impl_,
+        )
+    }
+}
+
+impl Eq for GuardCondition {}
+
+// SAFETY: rcl_guard_condition is the only member that doesn't implement Send, and it is designed to be accessed from other threads
+unsafe impl Send for rcl_guard_condition_t {}
+
+// SAFETY: The mutexes and atomic members ensure synchronized access to members, and the callback is reentrant
+unsafe impl Sync for GuardCondition {}
```

Comment:
Please remove this, it shouldn't be necessary. All members implement `Sync`.

Reply:
And the test confirms it!